### PR TITLE
Create one reference ScanSettings object by default

### PR DESCRIPTION
**Pull request by JAORMX**
to enhance the usability of the operator, this creates a ScanSettings
object called `default`, which contains reference values that could be
taken into use.

The default values are:
* raw storage allocation of 1Gi
* rotation policy of 3
* debugging turned off
* tolerations that allow scanning master nodes
* auto applying remediations is off
* the scans will run every day at 1am

This way, the only thing that an admin would need to do is create a
scanSettingBinding object to start running scans.

### CODE REVIEW

File: cmd/manager/operator.go [hunk 1]
```diff
@@ -241,6 +249,45 @@ func ensureDefaultProfileBundles(ctx context.Context, crclient client.Client, na
...
+	var lastErr error
+	for _, ns := range namespaceList {
+		pb := &compv1alpha1.ScanSetting{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      defaultScanSettingsName,
+				Namespace: ns,
+			},
+			ComplianceScanSettings: compv1alpha1.ComplianceScanSettings{
+				Debug: false,
+				RawResultStorage: compv1alpha1.RawResultStorageSettings{
```
Comment (jhrozek) [inline cmd/manager/operator.go:262]: I like the general idea of creating the default object, but do we need to explicitly mention the defaults? It appears that the defaults are applied by the respective controllers when the objects are already created. And should we ever change the defaults, the example object wouldn't change unless we also update it, unlike the defaults applied when creating the CRs.

Comment (JAORMX) [reply cmd/manager/operator.go:262]: We probably don't need to specify defaults for the default object. Let's do this, I'll only leave the defaults that come from a constant.

**Comment by jhrozek** [comment]
I have two remarks, not hard nacks, but I'm interested in what you think
1) There seem to be a bug somewhere, I'm sure it is not related to your patch, but I think we should file an issue and try to get to the root of it. The reason some of the scans fail is:
```
{"level":"info","ts":1597125319.7650719,"logger":"scanctrl","msg":"Node scan error","node.Name":"ip-10-0-166-55.us-west-2.compute.internal","errMsg":"the ConfigMap 'openscap-pod-fd08fa28d8a49f1f9c75c264db59ab60ebc5e38a' annotation was missing the result key"}
```
but, the configmap DOES have the result key:
```
"kind": "ConfigMap",
            "metadata": {
                "annotations": {
                    "compliance-remediations/processed": "",
                    "compliance.openshift.io/scan-error-msg": "",
                    "compliance.openshift.io/scan-result": "COMPLIANT",
                    "openscap-scan-result/node": "ip-10-0-166-55.us-west-2.compute.internal"
                },
                "creationTimestamp": "2020-08-11T05:53:11Z",
                "labels": {
                    "compliance-scan": "test-scan-w-missing-tailoring-cm"
                },
                "managedFields": [
                    {
                        "apiVersion": "v1",
                        "fieldsType": "FieldsV1",
                        "fieldsV1": {
                            "f:data": {
                                ".": {},
                                "f:exit-code": {},
                                "f:results": {}
                            },
                            "f:metadata": {
                                "f:annotations": {
                                    ".": {},
                                    "f:compliance-remediations/processed": {},
                                    "f:compliance.openshift.io/scan-error-msg": {},
                                    "f:compliance.openshift.io/scan-result": {},
                                    "f:openscap-scan-result/node": {}
                                },
                                "f:labels": {
                                    ".": {},
                                    "f:compliance-scan": {}
                                }
                            }
                        },
                        "manager": "compliance-operator",
                        "operation": "Update",
                        "time": "2020-08-11T05:55:17Z"
                    }
                ],
                "name": "openscap-pod-fd08fa28d8a49f1f9c75c264db59ab60ebc5e38a",
```
and the aggregator pod says that it had annotated the objects:
```
{"level":"info","ts":1597125316.3485558,"logger":"cmd","msg":"Annotating ConfigMaps"}
```
Looking at the timestamps, the annotating configMaps message was emitted at 1597125316.3485558 (=Tue 11 Aug 2020 05:55:16 AM UTC) and the check for the annotations happened at 1597125319.7650719 (=Tue 11 Aug 2020 05:55:19 AM UTC), so three seconds later.
Could it be that the informer is returning obsolete data? I'm really not sure how to handle this, but at the same time it's not the first time I've seen this issue either so I wouldn't like to just ignore it..

2) It still seems that the `RawResultStorage` defaults are not necessary

**Comment by JAORMX** [comment]
> I have two remarks, not hard nacks, but I'm interested in what you think
> 
>     1. There seem to be a bug somewhere, I'm sure it is not related to your patch, but I think we should file an issue and try to get to the root of it. The reason some of the scans fail is:
> 
> 
> ```
> {"level":"info","ts":1597125319.7650719,"logger":"scanctrl","msg":"Node scan error","node.Name":"ip-10-0-166-55.us-west-2.compute.internal","errMsg":"the ConfigMap 'openscap-pod-fd08fa28d8a49f1f9c75c264db59ab60ebc5e38a' annotation was missing the result key"}
> ```
> 
> but, the configmap DOES have the result key:
> 
> ```
> "kind": "ConfigMap",
>             "metadata": {
>                 "annotations": {
>                     "compliance-remediations/processed": "",
>                     "compliance.openshift.io/scan-error-msg": "",
>                     "compliance.openshift.io/scan-result": "COMPLIANT",
>                     "openscap-scan-result/node": "ip-10-0-166-55.us-west-2.compute.internal"
>                 },
>                 "creationTimestamp": "2020-08-11T05:53:11Z",
>                 "labels": {
>                     "compliance-scan": "test-scan-w-missing-tailoring-cm"
>                 },
>                 "managedFields": [
>                     {
>                         "apiVersion": "v1",
>                         "fieldsType": "FieldsV1",
>                         "fieldsV1": {
>                             "f:data": {
>                                 ".": {},
>                                 "f:exit-code": {},
>                                 "f:results": {}
>                             },
>                             "f:metadata": {
>                                 "f:annotations": {
>                                     ".": {},
>                                     "f:compliance-remediations/processed": {},
>                                     "f:compliance.openshift.io/scan-error-msg": {},
>                                     "f:compliance.openshift.io/scan-result": {},
>                                     "f:openscap-scan-result/node": {}
>                                 },
>                                 "f:labels": {
>                                     ".": {},
>                                     "f:compliance-scan": {}
>                                 }
>                             }
>                         },
>                         "manager": "compliance-operator",
>                         "operation": "Update",
>                         "time": "2020-08-11T05:55:17Z"
>                     }
>                 ],
>                 "name": "openscap-pod-fd08fa28d8a49f1f9c75c264db59ab60ebc5e38a",
> ```
> 
> and the aggregator pod says that it had annotated the objects:
> 
> ```
> {"level":"info","ts":1597125316.3485558,"logger":"cmd","msg":"Annotating ConfigMaps"}
> ```
> 
> Looking at the timestamps, the annotating configMaps message was emitted at 1597125316.3485558 (=Tue 11 Aug 2020 05:55:16 AM UTC) and the check for the annotations happened at 1597125319.7650719 (=Tue 11 Aug 2020 05:55:19 AM UTC), so three seconds later.
> Could it be that the informer is returning obsolete data? I'm really not sure how to handle this, but at the same time it's not the first time I've seen this issue either so I wouldn't like to just ignore it..

Right, it could be that it's obsolete data. Mostly we're using the controller-runtime's client, which uses a cache client to do the Get and List operations. These will hold objects for a specific amount of time and thus reduce the hits we do to the api server... but, it has the issue that some objects might be old. I guess this is not much of an issue for our main controllers, but the aggregator is having issues with this. So... we either move away from using the controller-runtime client and always hit the apiserver, or we address this limitation in the code.


> 
>     1. It still seems that the `RawResultStorage` defaults are not necessary

**Comment by jhrozek** [comment]
On Tue, Aug 11, 2020 at 05:14:25AM -0700, Juan Osorio Robles wrote:
> > Looking at the timestamps, the annotating configMaps message was emitted at 1597125316.3485558 (=Tue 11 Aug 2020 05:55:16 AM UTC) and the check for the annotations happened at 1597125319.7650719 (=Tue 11 Aug 2020 05:55:19 AM UTC), so three seconds later.
> > Could it be that the informer is returning obsolete data? I'm really not sure how to handle this, but at the same time it's not the first time I've seen this issue either so I wouldn't like to just ignore it..
> 
> Right, it could be that it's obsolete data. Mostly we're using the
> controller-runtime's client, which uses a cache client to do the Get and
> List operations. These will hold objects for a specific amount of time and
> thus reduce the hits we do to the api server... but, it has the issue that
> some objects might be old. I guess this is not much of an issue for our
> main controllers, but the aggregator is having issues with this. So... we
> either move away from using the controller-runtime client and always hit
> the apiserver, or we address this limitation in the code.

What bothers me is that we haven't really seen an issue with the cache
so far except this one. And I know too little about the informers and when does the
cache gets invalidated or updated to tell if it's really the root cause.

But, looping again to see if that helps might be an easy patch. I'll
take a look at submitting some test PR that backs off and retries a
number of times to see if it helps.

> 
> 
> > 
> >     1. It still seems that the `RawResultStorage` defaults are not necessary

What do you think about this?

**Comment by JAORMX** [comment]
> On Tue, Aug 11, 2020 at 05:14:25AM -0700, Juan Osorio Robles wrote: > Looking at the timestamps, the annotating configMaps message was emitted at 1597125316.3485558 (=Tue 11 Aug 2020 05:55:16 AM UTC) and the check for the annotations happened at 1597125319.7650719 (=Tue 11 Aug 2020 05:55:19 AM UTC), so three seconds later. > Could it be that the informer is returning obsolete data? I'm really not sure how to handle this, but at the same time it's not the first time I've seen this issue either so I wouldn't like to just ignore it.. Right, it could be that it's obsolete data. Mostly we're using the controller-runtime's client, which uses a cache client to do the Get and List operations. These will hold objects for a specific amount of time and thus reduce the hits we do to the api server... but, it has the issue that some objects might be old. I guess this is not much of an issue for our main controllers, but the aggregator is having issues with this. So... we either move away from using the controller-runtime client and always hit the apiserver, or we address this limitation in the code.
> What bothers me is that we haven't really seen an issue with the cache so far except this one. And I know too little about the informers and when does the cache gets invalidated or updated to tell if it's really the root cause. But, looping again to see if that helps might be an easy patch. I'll take a look at submitting some test PR that backs off and retries a number of times to see if it helps.

That would be good, thanks!

> > > 1. It still seems that the `RawResultStorage` defaults are not necessary
> What do you think about this?

Done

**Pull request merged**